src/core/net/srp_server.cpp

```diff
@@ -1199,6 +1201,68 @@ Error Server::VerifySignature(const Dns::Ecdsa256KeyRecord &aKeyRecord,
     return error;
 }
 
+Error Server::ValidateServiceSubTypes(Host &aHost, const MessageMetadata &aMetadata)
+{
+    Error error = kErrorNone;
+    Host *existingHost;
+
+    // Verify that there is a matching base type service for all
+    // sub-type services in `aHost` (which is from the received
+    // and parsed SRP Update message).
+
+    for (const Service &service : aHost.GetServices())
+    {
+        if (service.IsSubType() && (aHost.FindBaseService(service.GetInstanceName()) == nullptr))
+        {
+#if OT_SHOULD_LOG_AT(OT_LOG_LEVEL_WARN)
+            char subLabel[Dns::Name::kMaxLabelSize];
+
+            IgnoreError(service.GetServiceSubTypeLabel(subLabel, sizeof(subLabel)));
+            LogWarn("Message contains instance %s with subtype %s without base type", service.GetInstanceName(),
+                    subLabel);
+#endif
+
+            ExitNow(error = kErrorParse);
+        }
+    }
+
+    // SRP server must treat the update instructions for a service type
+    // and all its sub-types as atomic, i.e., when a service and its
+    // sub-types are being updated, whatever information appears in the
+    // SRP Update is the entirety of information about that service and
+    // its sub-types. Any previously registered sub-type that does not
+    // appear in a new SRP Update, must be removed.
+    //
+    // We go though the list of registered services for the same host
+    // and if the base service is included in the new SRP Update
+    // message, we add any previously registered service sub-type that
+    // does not appear in new Update message as "deleted".
```

Comment:
@abtink  Is this updated after we have the initial support of sub-types? I remember we have discussed about how to better implement on the SRP server and agreed to have separate service entry for each sub-type and have them point to the same base service entry. One benefit is that we can easily update each sub-types independently. But as it's now bundling lifetime of all sub-types to the base service, it may be simpler to let the `Service` class have a member field holds the sub-type list. thoughts?

Reply:
@wgtdkp , yes. I think it did.

Here is what I recall: We added sub-type support (since matter wanted to use) in SRP. On client side we kept it simple in the sense that all sub-types are always included in SRP update (we considered it specific behavior of OT SRP client). On server side, we kept the implementation more flexible (allowing subtypes in different SRP updates).  That's why we used this model on server: `Service` being either base or sub-type and can share the same underlying `Description`. The idea at the time was to support sub-types in multiple SRP updates on server side, and see if we want to change/enhance it on client later as well.

Later on (as Ted also looked into implementing this), we noticed allowing sub-types in different SRP updated can create a lot of unnecessarily complex situations, so in the latest SRP draft the atomic treatment of sub-types is required explicitly -> https://datatracker.ietf.org/doc/html/draft-ietf-dnssd-srp-13#section-2.3.4

I knew SRP client was already doing this, but recently noticed/realized that we may need to add this on server side (which is the goal of this PR).

> But as it's now bundling lifetime of all sub-types to the base service, it may be simpler to let the Service class have a member field holds the sub-type list. thoughts?

Yes. With the atomic requirement, we can simplify the SRP server's data model for sure, but I guess that's a bigger change, we can tackle it later.

Reply:
+1 to simplifying in a later PR.